add support for number format internationalization

fbaligand opened the pull request:
This PR adds a new advanced setting named "format:language".
This setting allows to choose a language to format numbers, among all numeraljs supported languages.
Number format is about decimal delimiter, thousand delimiter and currency symbol.

This feature is particularly useful to format every language specific number formats (not only english).

Precision : given how numeraljs works, it could only be a global option, this can't be customized for each field. Up to me, not a problem since it would be weird to display on a same dashboard different numeral fields with different language formats.

Fixes #9228.
Partially fixes #8983.

timroes commented on src/core_plugins/kibana/ui_setting_defaults.js:
```diff
@@ -226,6 +226,12 @@ export function getUiSettingDefaults() {
+      value: 'en',
+      type: 'select',
+      options: ['be-nl', 'chs', 'cs', 'da-dk', 'de-ch', 'de', 'en', 'en-gb', 'es-ES', 'es', 'et', 'fi', 'fr-CA', 'fr-ch', 'fr', 'hu', 'it', 'ja', 'nl-nl', 'pl', 'pt-br', 'pt-pt', 'ru-UA', 'ru', 'sk', 'th', 'tr', 'uk-UA'],
+      description: '<a href="http://numeraljs.com/" target="_blank">numeral language</a>'
```
Could you please add [`rel="noopener"`](https://mathiasbynens.github.io/rel-noopener/) to this link.

timroes commented on src/core_plugins/kibana/common/field_formats/types/_numeral.js:
```diff
@@ -26,6 +26,15 @@ export class Numeral extends FieldFormat {
       }
 
       getParamDefaults() {
+        var language = this.getConfig('format:language');
```
`var` -> `const`

timroes commented on src/core_plugins/kibana/common/field_formats/types/_numeral.js:
```diff
@@ -26,6 +26,15 @@ export class Numeral extends FieldFormat {
 
       getParamDefaults() {
+        var language = this.getConfig('format:language');
+        if (language != numeral.language()) {
```
`!=` → `!==`

timroes commented on src/core_plugins/kibana/common/field_formats/types/_numeral.js:
```diff
@@ -26,6 +26,15 @@ export class Numeral extends FieldFormat {
+        var language = this.getConfig('format:language');
+        if (language != numeral.language()) {
+          if (language != 'en') {
+            var languageData = require('@elastic/numeral/languages/' + language);
```
Please use a template string here.

timroes commented on src/core_plugins/kibana/common/field_formats/types/_numeral.js:
```diff
@@ -26,6 +26,15 @@ export class Numeral extends FieldFormat {
       getParamDefaults() {
+        var language = this.getConfig('format:language');
+        if (language != numeral.language()) {
+          if (language != 'en') {
```
`!=` → `!==`

fbaligand commented:
@timroes 
Done !

kobelb commented on src/core_plugins/kibana/common/field_formats/types/_numeral.js:
```diff
@@ -26,6 +26,15 @@ export class Numeral extends FieldFormat {
       }
 
       getParamDefaults() {
+        const language = this.getConfig('format:language');
```
We really shouldn't be setting the `numeral` language here, this method is only meant to return the param defaults. This situation appears to be very similar to setting the timezone for `moment` which is being done in the `KibanaRootController` (here)[https://github.com/elastic/kibana/blob/master/src/core_plugins/kibana/public/kibana_root_controller.js#L9] but then we're also having to do so in other app controllers, [example here](https://github.com/elastic/kibana/blob/master/src/core_plugins/timelion/public/app.js#L71).

Perhaps @spalger can comment on a better place to put logic like this without these deficiencies.

Additionally, I have hesitations to setting the language for the global instance of numeral as this could have unintended side-effects outside of modifying the behavior of field formatters specifically.

fbaligand replied:
I agree it's probably not the best place to set this code. 
If you find a better place, I would be glad to hear it.
I set it there because this method is called at app loading beginning and is rarely called then.

Concerning your hesitations, currently, numeraljs allows only to set this setting globally, not locally for each formatter. Then, would it make sense to have several language formats on the same kibana app ?

spalger replied:
This thread is kind of merging with https://github.com/elastic/kibana/pull/14019#discussion_r139841833

fbaligand replied:
@spalger do you see a better place (method) to put the javascript code which changes numeral language ?

spalger replied:
I'd just put it in the _convert function, then adding per-field params is just about exposing the UI to do so.

fbaligand replied:
If I set it there, it would require static `import` of all languages.

kobelb commented on src/core_plugins/kibana/ui_setting_defaults.js:
```diff
@@ -226,6 +226,12 @@ export function getUiSettingDefaults() {
+    'format:language': {
+      value: 'en',
+      type: 'select',
+      options: ['be-nl', 'chs', 'cs', 'da-dk', 'de-ch', 'de', 'en', 'en-gb', 'es-ES', 'es', 'et', 'fi', 'fr-CA', 'fr-ch', 'fr', 'hu', 'it', 'ja', 'nl-nl', 'pl', 'pt-br', 'pt-pt', 'ru-UA', 'ru', 'sk', 'th', 'tr', 'uk-UA'],
```
It's a shame that we're hard-coding these, but I'm not seeing a way to access all possible languages...

fbaligand replied:
@kobelb 
I think exactly like you :)
I'm sad to hard-code these values, but I didn't another way to do that.
That said, this array is likely to change very rarely (maybe never)...

spalger commented on src/core_plugins/kibana/ui_setting_defaults.js:
```diff
@@ -226,6 +226,12 @@ export function getUiSettingDefaults() {
       value: '($0,0.[00])',
       description: 'Default <a href="http://numeraljs.com/" target="_blank">numeral format</a> for the "currency" format'
     },
+    'format:language': {
```
:thumbsup: to making this setting more specific. I'm also slightly concerned about how this is a global setting. Perhaps `format:number:defaultLanguage` will keep it in line with `format:number:defaultPattern` and leaves room for a per-field language setting should we want to add one down the road.

fbaligand replied:
Hi,

`format:number:defaultLanguage` looks a good idea. 
That said, few points about that :
- first, this setting will apply on every type format concerned by numeraljs, so : number, bytes, percent, and currency. Not only number.
- then, currently (and since a long time), numeraljs allows only to define a global static language, not one for each formatter.
- would it make sense to have a dashboard with different numeric fields formatted with different languages ?
- in the future, an interesting evolution could be to detect browser language ?

=> Given these points, what setting name do you choose ? :)

spalger replied:
> numeraljs allows only to define a global static language

ugh, bummer. Is it unrealistic to call numeral.language() before every conversion?

> would it make sense to have a dashboard with different numeric fields formatted with different languages?

The numeral language defines the currency symbol right? If so I think it makes sense to have a unique language per field if two fields represent different currencies.

spalger replied:
Looks like `numeral.language()` is just setting an internal value inside the library, so if we called it before calling `numeralInst.set().format(this.param('pattern'))` we could use a different language for each field.

spalger replied:
@fbaligand I would probably move forward with `format:number:defaultLanguage` in this pr and do the per-field values in another pr, another day.

kobelb replied:
If we do move forward with applying this language to all instances of `numeral`, it'd be more consistent to do so in the manner described in this [comment](https://github.com/elastic/kibana/pull/14019#discussion_r139674713)

fbaligand replied:
@spalger I agree with you, it's fine to have `defaultLanguage` here to be able to add per-field value later.
That said, in original numeraljs project (https://github.com/adamwdraper/Numeral-js), the last version use the "locale" term instead of "language" term.
So question is : do you prefer `format:number:defaultLanguage` or `format:number:defaultLocale`

spalger replied:
I'm not too concerned about updates in numeral; I don't think we'll ever upgrade because of it's breaking changes and how that would impact everyone who has saved formats already. It's more likely that we will do something like we have done with kuery and offer a choice between legacy numeral and some custom solution we could build in the future.

That said, I think that locale is more accurate description of the choice, so I would go with `format:number:defaultLocale`

fbaligand replied:
OK, for this new name, I will commit it.

fbaligand replied:
@spalger 
new name 'format:number:defaultLocale' is commited

fbaligand commented on src/core_plugins/kibana/common/field_formats/types/_numeral.js:
```diff
@@ -26,6 +26,15 @@ export class Numeral extends FieldFormat {
+        const language = this.getConfig('format:language');
+        if (language !== numeral.language()) {
+          if (language !== 'en') {
+            var languageData = require(`@elastic/numeral/languages/${language}`);
```
@kobelb 
Sorry for this mistalke. I just pushed the fix.

fbaligand commented on src/core_plugins/kibana/common/field_formats/types/_numeral.js:
```diff
@@ -26,6 +26,15 @@ export class Numeral extends FieldFormat {
+        const language = this.getConfig('format:language');
+        if (language !== numeral.language()) {
+          if (language !== 'en') {
+            const languageData = require(`@elastic/numeral/languages/${language}`);
```
@spalger @kobelb 
Do you see a better way to load language data ?
Ideally, I would like to load statically all languages data to store them in numeral language map.
I try `import from @elastic/numeral/languages';` but it doesn't work.

spalger replied:
@fbaligand feel free to add the ability to do so at https://github.com/elastic/numeral-js

fbaligand replied:
@spalger 
One option for that is to use v3.0.0 that is already released. This version is compatible with static `import`. We can change the package.json file to depend on this version.
Are you OK to depend on this version ? Or do you consider this major version implies so much breaking changes ?

fbaligand replied:
@spalger 
I just commited migration to @elastic/numeral v3.0.0
I had no particular problem, it works fine !
And now I can `import` successfully all languages using a static way.
And so, I moved the code to `_convert` method as you wish.

fbaligand replied:
@spalger 
What do you think about ?

spalger replied:
Sorry, I should have deprecated `@elastic/numeral@3.0.0` [months ago](https://github.com/elastic/kibana/issues/7543#issuecomment-288508387). Unfortunately the way that numeral transforms format strings in v3 is not backwards compatible with v2. This means that in order to upgrade to numeral v3 we would either have to track down all the differences between the two versions, or users to recreate/double-check all of their numeric field formatters. We're currently not planning to do either, so this isn't going to work. I've updated [elastic/numeral-js](https://github.com/elastic/numeral-js) to reflect the current state of the package and changed https://github.com/elastic/numeral-js/blob/kibana-fork/languages.js

The file now exports the languages as an array. each item has an `id`, `name`, and `lang` property.
  - `id`: the filename of the lang
  - `name`: the english description of the language, [extracted from each language file](https://github.com/elastic/numeral-js/blob/kibana-fork/languages/de.js#L3)
  - `lang`: the value that should be passed to `numeral.language(id, lang)`

The file doesn't automatically register the languages, maybe we want it to do that, but it should be a suitable replacement for the current languages import, right? To test it out install `@elastic/numeral` from github:

```sh
npm install https://github.com/elastic/numeral-js#kibana-fork
```

fbaligand replied:
OK.
Thanks for all your work to transform "languages.js" in a way it can be imported easily !
So I'm ready to cancel my upgrade to 3.0.0 and write the code to register all languages in `_numeral.js` using your new `languages.js` version.
Where's the best place to put it ? Line 29 (https://github.com/fbaligand/kibana/blob/f6d8f5e251cd1c2e8d0aafb1fa9a65d2150557c4/src/core_plugins/kibana/common/field_formats/types/_numeral.js#L29) ?

spalger commented:
The conversation is flattened in my view of the PR, but I responded [here](https://github.com/elastic/kibana/pull/14019#discussion_r141418295)

fbaligand commented:
@spalger 
As you, I answered to the "old" conversation [here](https://github.com/elastic/kibana/pull/14019#discussion_r141425011).
Next time, if you prefer that I write a new comment, tell me.

fbaligand commented:
@spalger 
OK, I just cancelled my upgrade to 3.0.0 and wrote the code to register at startup, all languages in _numeral.js using your new languages.js version.
=> what do you think about ?

spalger commented on src/core_plugins/kibana/ui_setting_defaults.js:
```diff
@@ -226,6 +226,12 @@ export function getUiSettingDefaults() {
+    'format:number:defaultLocale': {
+      value: 'en',
+      type: 'select',
+      options: ['be-nl', 'chs', 'cs', 'da-dk', 'de-ch', 'de', 'en', 'en-gb', 'es-ES', 'es', 'et', 'fi', 'fr-CA', 'fr-ch', 'fr', 'hu', 'it', 'ja', 'nl-nl', 'pl', 'pt-br', 'pt-pt', 'ru-UA', 'ru', 'sk', 'th', 'tr', 'uk-UA'],
```
We should be able to replace this array with the ids from `@elastic/numeral/languages` right?

fbaligand replied:
Yes, it would be great to dynamically inject all language ids from "languages.js" array. 
That said, this configuration file seems to be completely static. 
=> do you know how to dynamically object values here ?

spalger replied:
You can require files at the top just like everything else, it's not actually static, just doesn't have any dependencies ***yet***.

fbaligand replied:
You're right, sorry for this remark. Even more, I just saw that there is already one dependency. 
I will fix this file to dynamically inject array from "languages.js".

spalger commented:
Jenkins, test this

fbaligand commented:
@spalger 
So that Jenkins build works, it should require a new release of "numeral" package, right ?

spalger commented:
Looks like jenkins build failed for linting reasons (run `npm run lint` locally) but you're right, I need to publish that.

spalger commented:
Just published `@elastic/numeral @ 2.3.0`

fbaligand commented:
Thanks for the release !
I will import it, and fix the lint issues detected by Jenkins

fbaligand commented:
@spalger
I just fixed jenkins lint issues, changed dependency to '@elastic/numeral' version 2.3.0, and injected dynamically languages ids into `format:number:defaultLocale` setting.

fbaligand commented:
Jenkins, test this

spalger commented:
Jenkins, test this

fbaligand commented:
@spalger 
Given Jenkins build logs, there are not `lint` errors anymore.

There are errors during tests. The main problem seems to be : 
```
22:29:22 Chrome 59.0.3071 (Linux 0.0.0) Range Agg formating formats bucket keys properly FAILED
22:29:22 	TypeError: this.getConfig is not a function
22:29:22 	    at Class._convert (http://localhost:5610/bundles/tests.bundle.js?shards=4&shard_num=2:420651:33)
22:29:22 	    at recurse (http://localhost:5610/bundles/tests.bundle.js?shards=4&shard_num=2:191577:25)
22:29:22 	    at FieldFormatFromConverter._convert (http://localhost:5610/bundles/tests.bundle.js?shards=4&shard_num=2:308585:46)
22:29:22 	    at recurse (http://localhost:5610/bundles/tests.bundle.js?shards=4&shard_num=2:191577:25)
22:29:22 	    at format (http://localhost:5610/bundles/tests.bundle.js?shards=4&shard_num=2:473843:37)
22:29:22 	    at Context.<anonymous> (http://localhost:5610/bundles/tests.bundle.js?shards=4&shard_num=2:473845:30)
```

I don't arrive to localize this test in Kibana sources, given the test label.
But it seems this error is because FieldFormatFromConverter calls directly `FieldFormat`, so it doesn't call _numeral `Class` constructor with `getConfig` parameter. That explains error above.

Honestly, I don't know how to fix this test. Any idea ?

fbaligand commented:
@spalger 
A way to work in all cases would be to put this code line :
`const defaultLocale = this.getConfig && this.getConfig('format:number:defaultLocale') || 'en';`

=> What do you think about ?

fbaligand commented:
@spalger 
I just published `const defaultLocale = this.getConfig && this.getConfig('format:number:defaultLocale') || 'en';` to fix jenkins tests.
=> could you launch jenkins tests ?

kobelb commented:
jenkins, test this

fbaligand commented:
@spalger @kobelb @nreese
Yes, jenkins build is OK !

=> Is this OK to merge the PR ?

kobelb commented:
I believe I brought this up earlier, and it might've already been addressed... are we not concerned that we're modifying the global numeral language from within the numeral field_formatter? It's conceivable that another Kibana application would be relying upon Numeral to do it's own formatting, and this would affect it's behavior as well. 

This could be incredibly confusing to the user in the following scenario. If there were two applications Kibana: App-1 relies upon numeraljs and NOT our field formatters; App-2 relies upon our field formatters and in turn numeraljs. When the user first opens Kibana they go directly to App-1, bypassing the field-formatter code that sets the default language. They then navigate to App-2 which is setting the global numeral language. Later if they go back to App-1 they see the number formats change.

fbaligand commented:
@kobelb 
If another app uses numeraljs directly (without Kibana core components), it probably depend on the official numeraljs, not @elastic/numeraljs.
What do you think about ?

kobelb commented:
@fbaligand we originally forked numeral-js not because we wanted a separate "instance" but because we were forced to make modifications to it; so the isolation that it provides in this situation doesn't assuage my concerns because in the future we could potentially switch to using the original project.

At this point, I feel like we have 2 ways forward:
1) Since numeral-js is forked, modify numeral-js so that we can set the language for a specific instance
2) Find a way to make it so that the numeral-js formatting truly is global to the application, and name the advanced setting accordingly. We've done something similar with moment timezones (with a few limitations) perhaps we could do the same here?

fbaligand commented:
I like your first idea.
Thus, the first step is to define a global default locale. The next one (not in this PR) would be to be able to define a locale "per field". So, in this case, clearly not a global locale.

Actually, @spalger made a new @elastic/numeraljs release with a major change to support properly all languages initial load. Thus, since official numeraljs 3.0 is out, some breaking changes have been done, so @elastic/numeraljs and official one have now taken 2 different ways (branches), and I'm very not confident that one day, these 2 projects get merged.

fbaligand commented:
@kobelb 
I just commited a change that restore previous locale just after conversion is done.
Doing that, it should answer your requested changes, using a simple way.

=> Could you tell me if it answers your requested changes ? And if so, launch jenkins tests ?

fbaligand commented:
@kobelb @spalger
Could you review my last commit ?

kobelb commented:
@fbaligand I can't think of why the approach that you've taken wouldn't work, but it does strike me as rather "hacky"... Let's see if it passes all the tests and go from there.

kobelb commented:
jenkins, test this

fbaligand commented:
@kobelb 
Sorry for these stupid lint issues...
I just fixed them. => could you launch Jenkins tests ?

kobelb commented:
Jenkins, test this

fbaligand commented:
@kobelb
Well, it seems that Jenkins build failed for technical reasons, independent from Kibana code. It seems it doesn't like my git commit message ...
What do you think?

fbaligand commented:
@kobelb 
As jenkins build seems to have failed because of the git commit itself (not the code) and this commit was directly via github UI with my smartphone (which is special...), I just replaced my previous commit with a new one from my laptop (more classic way...).

=> Could you launch jenkins tests ?

kobelb commented:
Jenkins, test this

fbaligand commented:
@kobelb 
OK, so my previous commit, there is not anymore "PGP signature/git commit read" error.
The problem is now further : ua-parser-js license is not OK.
I presume it is a new build check that require to be up-to-date with master.

=> I rebase my branch to upstream/master.
=> Could you launch jenkins tests ?

elasticmachine commented:
Can one of the admins verify this patch?

kobelb commented:
Apologies for the accidental close, clicked the wrong button 🥇

fbaligand commented:
@kobelb 
Yeah ! Jenkins build is successful !!

PS: no problem for accidental issue close :)

fbaligand commented:
@kobelb @spalger 
So now that jenkins build is successful, is this PR OK for you ?

fbaligand commented:
@kobelb @spalger
could you review a last time the PR ?
I hope it is "mergeable" now :)

fbaligand commented:
@kobelb @spalger
Could you review this PR ?
I've got no feedback for 3 weeks :(

kobelb commented:
I'm really not a fan of the current work-around, and would much prefer a proper solution where we create separate instances of numeral to reduce the risk of side-effects; however, @spalger is the resident expert on numeral-js so I'll defer to his judgement.

fbaligand commented:
Hi @kobelb,

Thank you for your feedback.
Actually, I looked at @elastic/numeral code to see what should be done so that "language" info is stored at numeral instance level (and not statically). And honesty, this is really impacting : numerous methods that use "language" info are static. 
So changing this would be clearly a breaking change.

@spalger what do you think about it ?

spalger commented:
I don't like the workaround either @kobelb, but I think it's just the cost of reusing existing code. Until we are ready to bring numeral into the project and make serious breaking changes to it this seems like our best option.

spalger commented:
jenkins, test this

kobelb commented:
@fbaligand mind merging in `elastic:master` to resolve the conflicts?

spalger commented:
@fbaligand I took care of merging master which should also take care of the jenkins failure

spalger commented:
jenkins, test this

fbaligand commented:
@spalger 
Thanks for the merge !

But it seems you make a little mistake in the merge.
Since line 45, up to me, the code should be : 
```
      const formatted = numeralInst.set(val).format(this.param('pattern'));

      numeral.language(previousLocale);

      return opts.afterConvert
        ? opts.afterConvert.call(this, formatted)
        : formatted;
```

fbaligand commented:
Actually, I rebased my PR (in the same time than you merged it) and squashed all my commits into only one.
=> Do you want I push (--force) the result ?

fbaligand commented:
OK, so Jenkins build failed apparently because this selenium test failed :
```
fail: "management scripted fields creating and using Painless numeric scripted fields should filter by scripted field value in Discover"
tryForTime timeout: Error: expected '386' to equal '31'
```

Looks like a broken test not related with this PR, no ?

spalger commented:
Ah, sorry about that @fbaligand go ahead and keep your version.

Can you try running the tests locally and see if they fail?

In one terminal tab run:
```sh
npm run test:ui:server
```

Once the Kibana server has started up in the first terminal run this in a second tab:
```sh
node scripts/functional_test_runner --grep "management scripted fields"
```

That will run the tests in test/functional/apps/management/_scripted_fields.js and might give you an indication if this is caused be the number formatters changing (a possibility if you ask me)

fbaligand commented:
@spalger 
I just pushed my rebased code, which fixes the removed code "opts.afterConvert" from your merge.
Maybe it was the cause of the failed selenium test ?
=> Could you run jenkins tests to see if build is fixed by my last push ?

Now, I have to go to sleep ;)

spalger commented:
Okay, that failure wasn't because of your code, for sure. One more jenkins run should do it

spalger commented:
jenkins test this

fbaligand commented:
@kobelb @spalger 
Yeah! Jenkins build is successful!

=> so now, what's the next step?

fbaligand commented:
Thanks @kobelb for the approval !

fbaligand commented:
@spalger @kobelb 
So is it OK to merge the PR ?

kobelb commented:
@fbaligand yup! I can merge this a little later today. Thanks for working through this with us and your patience.

fbaligand commented:
Great !
It was a pleasure to work with all of you on this PR !

fbaligand commented:
@spalger 
Thanks for the merge !!
Very happy that this feature is now merged and soon part of Kibana 6.2 !

fbaligand commented:
Is it too late to merge it in Kibana 6.1 ?

nreese commented:
> Is it too late to merge it in Kibana 6.1 ?

Yes

fbaligand commented:
Ok. Thanks for the answer @nreese.

spalger commented:
6.x/6.2: 91777d2

leeway23 commented:
Hi all, 

would anyone know from which Kibana version a problem previously discussed in (and later merged with this discussion) : 
https://github.com/elastic/kibana/issues/7543
is going to be fixed ? 

Thanks & Cheers, J.

MarcelZOI commented:
Hey, i currently looking how to change the number format of a column in visualize to german format. what is the current state? Is there a possibility to achive this?
Greetings

Marcel

fbaligand commented:
Hi @MarcelZOI 

As you can see in this PR tags, this feature is available since Kibana 6.2.0.

Important point : this setting is a Kibana advanced setting, so it is global. You currently can't have a custom locale per field.
